src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -339,41 +358,79 @@ private void ForceAuthentication(bool receiveFirst, byte[] buffer)
             }
         }
 
-        internal async Task ForceAuthenticationAsync(bool receiveFirst, byte[] buffer, CancellationToken cancellationToken)
+        private async Task ForceAuthenticationAsync<TReadAdapter>(TReadAdapter readAdapter, bool receiveFirst, byte[] inputData)
+            where TReadAdapter : ISslReadAdapter
         {
             _framing = Framing.Unknown;
-            ProtocolToken message;
-            SslReadAsync adapter = new SslReadAsync(this, cancellationToken);
+            ProtocolToken message = default;
+            bool isSync = readAdapter is SslReadSync;
 
             if (!receiveFirst)
             {
-                message = _context.NextMessage(buffer, 0, (buffer == null ? 0 : buffer.Length));
+                if (inputData?.Length > 0)
+                {
+                    _framing = DetectFraming(inputData);
+                }
+
+                message = _context.NextMessage(inputData);
+                if (message.Size > 0)
+                {
+                    // If there is message send it out even if call failed. It may contain TLS Alert.
+                    if (isSync)
+                    {
+                        InnerStream.Write(message.Payload);
+                    }
+                    else
+                    {
+                        await InnerStream.WriteAsync(message.Payload, readAdapter.CancellationToken).ConfigureAwait(false);
+                    }
```

Comment:
I don't understand why this if/else block is necessary.  Why can't you just use the adapter to do the write?  Is the issue that you have a read adapter but you need a write adapter?  If so, you can't pass a write adapter in here?

Maybe the read and write adapters should be combined, so that we just have an SslSyncAdapter and an SslAsyncAdapter, each of which has both the read and write methods on it?

Reply:
yes, combing them together would do the trick.